dcicutils/aws_utils.py

```diff
@@ -0,0 +1,261 @@
+""" Functions to support general aws operations
+
+    Function names should generally follow a convention to include a short string
+    to indicate the service that they are designed to operate with eg. 's3' for AWS s3.
+"""
+import boto3
+import logging
+import mimetypes
+
+from botocore.exceptions import ClientError
+# from .exceptions import ()
+from .s3_utils import s3Utils
+
+###########################
+# Config
+###########################
+logging.basicConfig()
+logger = logging.getLogger(__name__)
+
+
+def s3_bucket_head(*, bucket_name, s3=None):
+    """ Gets head info for a bucket if it exists
+
+    :param bucket_name: name of the bucket - string
+    :param s3: AWS s3 client
+    :return: dict: head response or None
+    """
+    try:
+        s3 = s3 or s3Utils().s3
+        info = s3.head_bucket(Bucket=bucket_name)
+        return info
+    except ClientError:
+        return None
+
+
+def s3_bucket_exists(*, bucket_name, s3=None):
+    """ Does a bucket exist?
+
+    :param bucket_name: name of the bucket - string
+    :param s3: AWS s3 client
+    :return: boolean - True if exists, False if not
+    """
+    return bool(s3_bucket_head(bucket_name=bucket_name, s3=s3))
+
+
+def s3_bucket_object_count(bucket_name):
+    """ Number of objects in the given s3 bucket
+
+    NB: this works with locally stored credentials - not sure
+    if something needs to be added or if it is possible to create
+    a resource from a client or provide credentials if not locally stored
+
+    :param bucket_name: name of the bucket - string
+    :return: int - number of objects in bucket
+    """
+    bucket = boto3.resource('s3').Bucket(bucket_name)
+    # get only head of objects so we can count them
+    return sum(1 for _ in bucket.objects.all())
+
+
+def s3_object_head(*, object_key, bucket_name, s3=None):
+    """ Gets head info for a object if it exists in provided bucket
+
+    :param object_key: key for the object - string
+    :param bucket_name: name of the bucket to check for the object - string
+    :param s3: AWS s3 client
+    :return: dict - head response or None
+    """
+    try:
+        s3 = s3 or s3Utils().s3
+        info = s3.head_object(Bucket=bucket_name, Key=object_key)
+        return info
+    except ClientError:
+        return None
+
+
+def s3_object_exists(*, object_key, bucket_name, s3=None):
+    """ Does an object exist in the given bucket?
+
+    :param object_key: key for the object - string
+    :param bucket_name: name of the bucket - string
+    :param s3: AWS s3 client
+    :return: boolean - True if exists, False if not
+    """
+    return bool(s3_object_head(object_key=object_key, bucket_name=bucket_name, s3=s3))
+
+
+def s3_put_object(*, object_key, obj, bucket_name, acl=None, s3=None):
```

Comment:
You should handle the `kms_key_id` case here as well, despite it not being used in Fourfront